plugins/kubernetes/app/models/watchers/deploy_watcher.rb

```diff
@@ -20,18 +21,44 @@ def watch
 
     def handle_update(topic, data)
       release_doc = release_doc_from_rc_name(topic)
-      pod_event = Events::PodEvent.new(data)
-      return error('invalid k8s pod event') unless pod_event.valid?
-      update_replica_count(release_doc, pod_event)
-      send_event(role: release_doc.kubernetes_role.name,
-                 deploy_group: release_doc.deploy_group.name,
-                 target_replicas: release_doc.replica_target,
-                 live_replicas: release_doc.replicas_live)
-      end_deploy if deploy_finished?
+      updates_found = if data.object.kind == 'Event'
+                        handle_event_update(release_doc)
+                      else
+                        handle_pod_update(release_doc, data)
+                      end
+      if updates_found
+        send_event(role: release_doc.kubernetes_role.name,
+                   deploy_group: release_doc.deploy_group.name,
+                   target_replicas: release_doc.replica_target,
+                   live_replicas: release_doc.replicas_live,
+                   failed: release_doc.failed?)
+        end_deploy if deploy_finished?
+      end
     end
 
     private
 
+    def handle_event_update(release_doc)
+      if release_doc.failed?
+        false # no update needed, the release has already failed
+      else
+        release_doc.update_attribute(:status, :failed)
```

Comment:
What do you think of refactoring this logic to the `ReleaseDoc`, so that only the 'ReleaseDoc' knows about the possible values for the status?

Then, this would look just like:
```
release_doc.fail!
```